OnkyoAVTCP: Add source selection, detection via NRIQSTN

dwmw2 opened the pull request:
Now it detects what the names of all the sources are, and lets me choose them. Also has support for 4 zones and automatically enables the ones that exist in the detected hardware.

Obsoletes #1958 and #1960

dwmw2 committed:
Stop limiting selectors to ten choices

dwmw2 committed:
Split out CSQLHelper::FormatDeviceOptions()

It's painful to have to create the device first, then find its ID, then
call CSQLHelper::SetDeviceOptions(). Just give me the string to insert
into the Options field in advance...

dwmw2 committed:
OnkyoAVTCP: Add source selection, detection via NRIQSTN

(Merged subsequent MSVC compatibility fixes)

dwmw2 commented:
```
..\hardware\OnkyoAVTCP.cpp(192): error C3480: 'ID_MVL': a lambda capture variable must be from an enclosing function scope [C:\projects\domoticz\msbuild\domoticz.vcxproj]
..\hardware\OnkyoAVTCP.cpp(192): error C3260: '=': skipping unexpected token(s) before lambda body [C:\projects\domoticz\msbuild\domoticz.vcxproj]
..\hardware\OnkyoAVTCP.cpp(192): error C2143: syntax error: missing ';' before '}' [C:\projects\domoticz\msbuild\domoticz.vcxproj]
```

FFS. Is that just because Microsoft haven't yet managed to implement a C standard from the last century? What is the MSVC-safe way to do explicit initialisers? This is the offending code:
```
static struct {
	…
} switch_types[] = {
	[ID_MVL] = { "MVL", "AMT", "Master volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	[ID_ZVL] = { "ZVL", "ZMT", "Zone 2 volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	…
};
```

dwmw2 commented:
I killed the explicit initialisation. It was all in order with no gaps anyway; I just wanted to be extra safe.

gizmocuz commented on hardware/OnkyoAVTCP.cpp:
```diff
@@ -305,26 +356,157 @@ void OnkyoAVTCP::ReceiveSwitchMsg(const char *pData, int Len, bool muting, int I
-			if (result.size() > 0) {
-				std::string Idx = result[0][0];
-				m_sql.SetDeviceOptions(atoi(Idx.c_str()), m_sql.BuildDeviceOptions(switch_types[ID].options, false));
+	sDecodeRXMessage(this, (const unsigned char *)&gswitch, switch_types[ID].name, 255);
```
Please use the default SendSwitch function

dwmw2 replied:
Ah yes, I think I can so that now that I've made the device creation saner. Will fix.

dwmw2 replied:
Hm, maybe not... that sends a lighting2 message, and this is a generalswitch. I also do still have special magic to do, to look up the target level for the selector from LevelActions. I'm not quite sure I see how to make it use SendSwitch, or at least I don't see that it simplifies anything. Am I missing something?

gizmocuz replied:
Does it matter is it is a lighting2 or general switch ? Or are you using a selector switch ?
There might be a SendSelectorSwitch option soon, till then please use you current method

dwmw2 replied:
I have selector, on/off and level switches. Currently they are all `pTypeGeneralSwitch` devices and handled generically, based on the following table:

```
	{ "MVL", "AMT", "Master volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	{ "ZVL", "ZMT", "Zone 2 volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	{ "VL3", "MT3", "Zone 3 volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	{ "VL4", "MT4", "Zone 4 volume", STYPE_Dimmer, sSwitchGeneralSwitch, 8, NULL },
	{ "PWR", NULL, "Master power", STYPE_OnOff, sSwitchGeneralSwitch, 5, NULL },
	{ "ZPW", NULL, "Zone 2 power", STYPE_OnOff, sSwitchGeneralSwitch, 5, NULL },
	{ "PW3", NULL, "Zone 3 power", STYPE_OnOff, sSwitchGeneralSwitch, 5, NULL },
	{ "PW4", NULL, "Zone 4 power", STYPE_OnOff, sSwitchGeneralSwitch, 5, NULL },
	{ "SLI", NULL, "Master selector", STYPE_Selector, sSwitchTypeSelector, 5, NULL },
	{ "SLZ", NULL, "Zone 2 selector", STYPE_Selector, sSwitchTypeSelector, 5, NULL },
	{ "SL2", NULL, "Zone 3 selector", STYPE_Selector, sSwitchTypeSelector, 5, NULL },
	{ "SL3", NULL, "Zone 4 selector", STYPE_Selector, sSwitchTypeSelector, 5, NULL },
	{ "HDO", NULL, "HDMI Output", STYPE_Selector, sSwitchTypeSelector, 5, "SelectorStyle:0;LevelNames:Off|Main|Sub|Main+Sub;LevelOffHidden:true;LevelActions:00|01|02|03" },
```

gizmocuz replied:
For now please keep using your direct calls. I think there will be a function to send/update selector switches shortly, maybe then we can replace these functions

gizmocuz commented on hardware/OnkyoAVTCP.cpp:
```diff
@@ -364,7 +546,7 @@ void OnkyoAVTCP::ParseData(const unsigned char *pData, int Len)
 		}
 		int data_size = static_cast<int>(ntohl(pkt->data_size));
-//		_log.Log(LOG_NORM, "ISCP at len %d size %d", Len, data_size);
+		_log.Log(LOG_NORM, "ISCP at len %d size %d", Len, data_size);
```
I think this is a debug log and should be removed?

dwmw2 replied:
Removed.

gizmocuz commented on hardware/OnkyoAVTCP.h:
```diff
@@ -3,6 +3,7 @@
 #include <iosfwd>
 #include "ASyncTCP.h"
 #include "DomoticzHardware.h"
+#include <map>
```
I see here a map include but no reference to a map usage in this header? If it is used in the .cpp, then please include this in the .cpp only

dwmw2 replied:
Yep, there *was* a map but I simplified things. Will fix it.

gizmocuz commented on main/RFXNames.cpp:
```diff
@@ -1963,7 +1963,7 @@ void GetSelectorSwitchStatuses(const std::map<std::string, std::string> & option
 		int i = 0;
 		std::stringstream ss;
-		for (itt = strarray.begin(); (itt != strarray.end()) && (i <= 100); ++itt) {
+		for (itt = strarray.begin(); (itt != strarray.end()); ++itt) {
```
In the feature please make two separate pull requests because this one targets two different issues.
If for any reason we need to revert this pull request, this patch will also be removed, or visa versa

dwmw2 replied:
I can do that right now if you prefer. The disadvantage is that then we have two separate pull requests that depend on each other and if you revert the other one, you break this one anyway.

gizmocuz replied:
Plz leave it in.. i've seen the other pull request, and both look fine!

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -7820,14 +7815,25 @@ bool CSQLHelper::SetDeviceOptions(const uint64_t idx, const std::map<std::string
 	}
-	if (options.empty()) {
+
+	if (optionsMap.size() == 0) {
```
this should be optionsMap.empty() please

dwmw2 replied:
Fixed.

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -7795,12 +7795,7 @@ std::map<std::string, std::string> CSQLHelper::GetDeviceOptions(const std::strin
-		return false;
-	}
-
+std::string CSQLHelper::FormatDeviceOptions(const std::map<std::string, std::string> & optionsMap) {
```
optionsMap could be declared const

dwmw2 replied:
Um... isn't it?
```const map<string, string> & optionsMap```

gizmocuz replied:
Whoops sorry..

dwmw2 commented:
I've addressed the feedback I could understand.
Not done: 
 • make optionsMap more const
 • use SendSwitch()

As far as I can tell, using SendSwitch() wouild require changing the type of the underlying devices to a lighting switch?

gizmocuz commented:
Thanks!
